src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_Hierarchy.cs

```diff
@@ -275,5 +275,257 @@ public static Task<IEnumerable<SymbolCallerInfo>> FindCallersAsync(ISymbol symbo
+
+            // Must contain non-nested named types.
+            Contract.ThrowIfFalse(equivalentTypesWithDifferingAssemblies.All(kvp => kvp.Key.ContainingType == null));
+            Contract.ThrowIfFalse(equivalentTypesWithDifferingAssemblies.All(kvp => kvp.Value.ContainingType == null));
+
+            var referencedAssemblies = new MultiDictionary<string, IAssemblySymbol>();
+            foreach (var assembly in compilation.GetReferencedAssemblySymbols())
+            {
+                referencedAssemblies.Add(assembly.Name, assembly);
+            }
+
+            int verifiedCount = 0;
+            foreach (var kvp in equivalentTypesWithDifferingAssemblies)
+            {
+                if (!verifiedKeys.Contains(kvp.Key))
+                {
+                    INamedTypeSymbol originalType, expectedForwardedType;
+                    if (isSearchSymbolCompilation)
+                    {
+                        originalType = kvp.Value.OriginalDefinition;
+                        expectedForwardedType = kvp.Key.OriginalDefinition;
+                    }
+                    else
+                    {
+                        originalType = kvp.Key.OriginalDefinition;
+                        expectedForwardedType = kvp.Value.OriginalDefinition;
+                    }
+
+                    foreach (var referencedAssembly in referencedAssemblies[originalType.ContainingAssembly.Name])
+                    {
+                        var fullyQualifiedTypeName = originalType.MetadataName;
+                        if (originalType.ContainingNamespace != null)
+                        {
+                            fullyQualifiedTypeName = originalType.ContainingNamespace.ToDisplayString(SymbolDisplayFormats.SignatureFormat) +
+                                "." + fullyQualifiedTypeName;
+                        }
+
+                        // Resolve forwarded type and verify that the types from different assembly are indeed equivalent.
+                        var forwardedType = referencedAssembly.ResolveForwardedType(fullyQualifiedTypeName);
+                        if (forwardedType == expectedForwardedType)
+                        {
+                            verifiedKeys.Add(kvp.Key);
+                            verifiedCount++;
+                        }
+                    }
+                }
+            }
+
+            return verifiedCount;
+        }
+
+        private static bool TryGetCompilation(
+            ISymbol symbol,
+            Solution solution,
+            out Compilation definingCompilation,
+            CancellationToken cancellationToken)
+        {
+            var definitionProject = solution.GetProject(symbol.ContainingAssembly, cancellationToken);
+            if (definitionProject == null)
+            {
+                definingCompilation = null;
+                return false;
+            }
+
+            // compilation from definition project must already exist.
```

Comment:
Why? What if symbol is from metadata and we get bit by that bug that took you and @dpoeschl a few days to unravel?